### ISSUE TITLE
Airbase offline and removed from Homekit Automation

### ISSUE THREAD

**Issue by sgraystar** [yenoiwesa/homebridge-daikin-airbase#18]
Thanks for the plugin, I have been using it successfully for some time.

However when power is lost it takes a long time for my Airbase controller to appear on the LAN, not sure why, perhaps due to the fact it is not officially supported on my old HVAC.

As a result Homekit removes the accessory from any automations, requiring manual intervention to recover automation rules.

I was expecting Homebridge/Homekit to recover when the Airbase eventually reappeared online, is there some kind of cleanup if the accessory is offline for too long?

**Comment by yenoiwesa** [comment]
Hi @sgraystar 👋 

It's funny you'd mention that problem today because I just experienced the same problem with my other homebridge plugin that maps my Somfy blinds to HomeKit.

Homebridge does keep an accessory cache, which is why the room allocation and the custom names and icons of your accessories remain once your restart the homebridge server. The problem however is that once your homebridge server restarts, the plugin's platform will look for the accessories on the network. If it does not manage to find them or contact them at that point, they will not be registered. In consequence, homebridge will just forget about them and clear their cache. The next time they are seen, they will appear as new accessories.

What I intend to do for the Somfy plugin is to keep a cache of previously discovered accessories on disk, that will remain until I manage to contact the Somfy server again (the source of truth).

However for the Daikin Airbase, it's a bit more tricky, because the discovery happens on the local network via a broadcast packet. It's hard to tell the difference between an accessory not being reachable yet, or one that just no longer exists. So what condition should I use to determine that this old accessory I have in cache is no longer relevant and should be purged and no longer registered?

What I am thinking is that I might keep the airbase cache forever, until the user removes the cache file from the disk (with instructions on how to do so in the README file). I'd only do the auto discovery when the platform starts if there is no accessory previously discovered in cache (otherwise I'd skip the auto discovery).

What do you think?

**Comment by TimofeyK** [comment]
@yenoiwesa 
Air-conditioners are pretty “static” devices, they are installed for decades. If an accessory is not reachable it’s most likely a connectivity issue rather than the controller is physically removed

If someone removes the Airbase controller this would mean removal of the plugin as well so you don’t need to implement this scenario

Only two things are likely to change during the life cycle if the controller: IP address of the controller and zone names

**Comment by TimofeyK** [comment]
I think it might be better logic: do auto-discovery or use fixed configured IP (current logic) and if no device is detected/reachable - use the last known configuration from the cache

**Comment by sgraystar** [comment]
There must be something 'special' with my setup, it is very unlikely the Airbase designers would not renew an IP asap after reboot. Under normal operation a delay of a few minutes before starting homebridge should be more than enough for everything to re-establish on the LAN.

If you decide to leave as-is I would understand.

As for generic use:
A persistent cache would cover accidental/temp scenarios where the controller is undiscoverable (maintenance, router snafu).
It is also possible to deploy more than one controller on the LAN, therefor possible to remove one controller and keep the rest.
If the firmware auto-updates it is also theoretically possible for the DaikinAPxxxxx ID to change format.
To keep the accessory would also require introduction of an 'offline' state, otherwise the accessory would remain in homekit & automations and there is nothing to show users why it does not work.
I cannot see how you could tell the difference between a controller that is not reachable and one that has failed or been removed, so a multi-day timer or manual intervention are possibly the only options for purging a persistent cache.

**Comment by yenoiwesa** [comment]
@sgraystar @TimofeyK hey guys I have made some pretty important refactoring to the plugin to allow two things:
- first, the plugin will now re-use the cached accessories from homebridge and re-hydrate them once homebridge starts, instead of just creating new accessories from scratch every time.
- second, the plugin will try to re-associate previously cached accessories to their airbase on the network using their SSID as unique identifier. If it fails to re-associate all of them, it will retry up to 10 times, with a an exponential delay starting from 5 seconds. If after these attempts some cached accessories are still not associated to an airbase found on the network, it will unregister these accessories from homekit.

All in all, that means it will make its last attempt about 85 minutes after the server has started, which I think is fair. I could adjust these values if you think that won't be enough.

Because there are so many changes, could you guys test it a bit (and especially @TimofeyK since I still don't have a zone controller at home 😄) and let me know if everything behaves fine?

You can install the unreleased version via:
```
npm install -g yenoiwesa/homebridge-daikin-airbase#93b8dea8cfd98ea5fc9ebe03a2e14a462f18fcec
```

(Only add the `-g` if you install your homebridge plugins globally)

If everything is fine I will release it!

Note: I have upgraded the homebridge dependency to `v1.0.0` as the minimum.

**Comment by TimofeyK** [comment]
@yenoiwesa 
On first restart after update all aircon accessories were recreated so disappeared from Configured room and all HomeKit automations / button actions. Is that a side effect of the refactoring and expected?

On the second restart Homebridge won’t run so I had to revert to the previous release:
```
[8/31/2020, 7:03:47 PM] [My Daikin Airbase Hub] Restoring cached accessory Daikin
[8/31/2020, 7:03:47 PM] [My Daikin Airbase Hub] Restoring cached accessory Daikin Zones
[8/31/2020, 7:03:47 PM] [My Daikin Airbase Hub] Failed to restore cached accessory Daikin Zones TypeError: this.context.airbase.zoneNames is not iterable
    at new ZoneControl (/usr/lib/node_modules/homebridge-daikin-airbase/src/accessories/zone-control.js:8:53)
    at DaikinAirbasePlatform.configureAccessory (/usr/lib/node_modules/homebridge-daikin-airbase/src/daikin-airbase-platform.js:42:31)
    at /usr/lib/node_modules/homebridge/src/server.ts:305:25
    at Array.filter (<anonymous>)
    at Server.restoreCachedPlatformAccessories (/usr/lib/node_modules/homebridge/src/server.ts:268:69)
    at Server.start (/usr/lib/node_modules/homebridge/src/server.ts:156:10)
    at cli (/usr/lib/node_modules/homebridge/src/cli.ts:80:10)
    at Object.<anonymous> (/usr/lib/node_modules/homebridge/bin/homebridge:17:22)
    at Module._compile (internal/modules/cjs/loader.js:1137:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1157:10)
[8/31/2020, 7:03:47 PM] [My Daikin Airbase Hub] Restoring cached accessory Daikin Zones
[8/31/2020, 7:03:47 PM] [My Daikin Airbase Hub] Failed to restore cached accessory Daikin Zones TypeError: this.context.airbase.zoneNames is not iterable
    at new ZoneControl (/usr/lib/node_modules/homebridge-daikin-airbase/src/accessories/zone-control.js:8:53)
    at DaikinAirbasePlatform.configureAccessory (/usr/lib/node_modules/homebridge-daikin-airbase/src/daikin-airbase-platform.js:42:31)
    at /usr/lib/node_modules/homebridge/src/server.ts:305:25
    at Array.filter (<anonymous>)
    at Server.restoreCachedPlatformAccessories (/usr/lib/node_modules/homebridge/src/server.ts:268:69)
    at Server.start (/usr/lib/node_modules/homebridge/src/server.ts:156:10)
    at cli (/usr/lib/node_modules/homebridge/src/cli.ts:80:10)
    at Object.<anonymous> (/usr/lib/node_modules/homebridge/bin/homebridge:17:22)
    at Module._compile (internal/modules/cjs/loader.js:1137:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1157:10)
[8/31/2020, 7:03:47 PM] Error: Cannot add a bridged Accessory with the same UUID as another bridged Accessory: 8cb58277-289a-4326-a8cf-9f9855ee33e0
    at Bridge.Accessory._this.addBridgedAccessory (/usr/lib/node_modules/homebridge/node_modules/hap-nodejs/src/lib/Accessory.ts:417:15)
    at /usr/lib/node_modules/homebridge/src/server.ts:308:19
    at Array.filter (<anonymous>)
    at Server.restoreCachedPlatformAccessories (/usr/lib/node_modules/homebridge/src/server.ts:268:69)
    at Server.start (/usr/lib/node_modules/homebridge/src/server.ts:156:10)
    at cli (/usr/lib/node_modules/homebridge/src/cli.ts:80:10)
    at Object.<anonymous> (/usr/lib/node_modules/homebridge/bin/homebridge:17:22)
    at Module._compile (internal/modules/cjs/loader.js:1137:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1157:10)
    at Module.load (internal/modules/cjs/loader.js:985:32)
[8/31/2020, 7:03:47 PM] Got SIGTERM, shutting down Homebridge...
[8/31/2020, 7:03:52 PM] [HB Supervisor] Homebridge Process Ended. Code: 143, Signal: null
```

**Comment by TimofeyK** [comment]
Forgot to mention: I am running Homebridge 1.2.0-beta.35
Node v12.18.3

**Comment by TimofeyK** [comment]
BTW, I don’t think I ever had my aircon accessories to be removed from HomeKit because the controller was not reachable. Probably I’ve never restarted Homebridge while the controller was offline.

I am still not convinced that the plugin should ever drop accessories and 85 minutes might be not enough for fancoil maintenance (it will be powered off so the controller will be powered off too). I am not sure if I understood what would happen if the controller is unreachable in the network

Will the refactored logic pickup new zone names if they are changed on the physical controller or the Daikin app?

**Comment by yenoiwesa** [comment]
😞 

> On first restart after update all aircon accessories were recreated so disappeared from Configured room and all HomeKit automations / button actions. Is that a side effect of the refactoring and expected?

Yes that is expected because I went from a static platform to a dynamic platform so accessories are registered differently.
Sorry for the pain of having to recreate your automations, I should have warned you before :-/

Regarding the issue with the cached accessory on the first start, that is probably because the homebridge context does not know how to serialise a javascript Set. I have now transformed zone names back into an array.

```
npm install -g yenoiwesa/homebridge-daikin-airbase#8f026c59aa928f95ba4296de2bc3aef6660e671a
```

> BTW, I don’t think I ever had my aircon accessories to be removed from HomeKit because the controller was not reachable. Probably I’ve never restarted Homebridge while the controller was offline.

It must be because of that yes. However that could happen during a power outage.

> I am still not convinced that the plugin should ever drop accessories and 85 minutes might be not enough for fancoil maintenance (it will be powered off so the controller will be powered off too). I am not sure if I understood what would happen if the controller is unreachable in the network

I am only un-registering accessories if they are unreachable after 85 minutes **WHEN** homebridge is restarted. It would really be bad luck to have homebridge restart at the beginning of an aircon maintenance. I could also never drop accessories but then there would be no other way for users than erase entirely the entire homebridge persistence cache (not just for the Daikin Airbase). That doesn't sound too acceptable.

> Will the refactored logic pickup new zone names if they are changed on the physical controller or the Daikin app?

It will, however upon two restarts only, which is admittedly crap. The first restart will use the cached values that come first, but update the cache, the second one will pick up the updated cache.
I can improve that behaviour and automatically add new zone switches and remove unused ones, but it's additional code and I am not too sure if it's worth doing. What do you think?

**Comment by TimofeyK** [comment]
- The exception has gone. Tried 3 restarts
- Zones support appears to work as before

Two restarts on zone name changes are acceptable, I think, if this is documented. I’d rather avoid rewriting working code (that you are not able to test) for that minor issue

What I actually didn’t test:
- Automatic discovery - I use fixes IP address in the config to shorten startup time and my controller uses static IP lease
- Putting the controller offline (especially for 85 minutes) :)
- Changing zone names

**Comment by TimofeyK** [comment]
Tested configuring wring IP to simulate offline device:
```
[8/31/2020, 8:18:32 PM] Homebridge is running on port 51365.
[8/31/2020, 8:18:42 PM] [My Daikin Airbase Hub] TypeError: Cannot read property 'ok' of undefined
    at DaikinAircon.sendRequest (/usr/lib/node_modules/homebridge-daikin-airbase/src/airbase-controller.js:128:23)
    at processTicksAndRejections (internal/process/task_queues.js:97:5)
[8/31/2020, 8:18:42 PM] [My Daikin Airbase Hub] Will retry to find missing airbase modules with SSIDs [ 'DaikinAP08350' ]
[8/31/2020, 8:18:56 PM] [My Daikin Airbase Hub] TypeError: Cannot read property 'ok' of undefined
    at DaikinAircon.sendRequest (/usr/lib/node_modules/homebridge-daikin-airbase/src/airbase-controller.js:128:23)
    at processTicksAndRejections (internal/process/task_queues.js:97:5)
[8/31/2020, 8:18:56 PM] [My Daikin Airbase Hub] Will retry to find missing airbase modules with SSIDs [ 'DaikinAP08350' ]
[8/31/2020, 8:19:10 PM] [My Daikin Airbase Hub] No airbase is associated to Daikin Zones
[8/31/2020, 8:19:10 PM] [My Daikin Airbase Hub] No airbase is associated to Daikin Zones
[8/31/2020, 8:19:10 PM] [My Daikin Airbase Hub] No airbase is associated to Daikin Zones
[8/31/2020, 8:19:10 PM] [My Daikin Airbase Hub] No airbase is associated to Daikin Zones
[8/31/2020, 8:19:10 PM] [My Daikin Airbase Hub] No airbase is associated to Daikin
[8/31/2020, 8:19:10 PM] [My Daikin Airbase Hub] No airbase is associated to Daikin
[8/31/2020, 8:19:10 PM] [My Daikin Airbase Hub] No airbase is associated to Daikin
[8/31/2020, 8:19:10 PM] [My Daikin Airbase Hub] No airbase is associated to Daikin
[8/31/2020, 8:19:10 PM] [My Daikin Airbase Hub] No airbase is associated to Daikin
[8/31/2020, 8:19:10 PM] [My Daikin Airbase Hub] No airbase is associated to Daikin
[8/31/2020, 8:19:10 PM] [My Daikin Airbase Hub] No airbase is associated to Daikin
[8/31/2020, 8:19:10 PM] [My Daikin Airbase Hub] No airbase is associated to Daikin
[8/31/2020, 8:19:10 PM] [My Daikin Airbase Hub] No airbase is associated to Daikin
[8/31/2020, 8:19:16 PM] [My Daikin Airbase Hub] TypeError: Cannot read property 'ok' of undefined
    at DaikinAircon.sendRequest (/usr/lib/node_modules/homebridge-daikin-airbase/src/airbase-controller.js:128:23)
    at processTicksAndRejections (internal/process/task_queues.js:97:5)
[8/31/2020, 8:19:16 PM] [My Daikin Airbase Hub] Will retry to find missing airbase modules with SSIDs [ 'DaikinAP08350' ]
```

After restoring the correct IP and restart accessories picked up

**Comment by yenoiwesa** [comment]
That's great, thanks a lot for your help with the testing @TimofeyK. I did similar testing with setting a wrong IP on a second restart, and testing the unregistering of orphan accessories by changing the number of retries to just one. I am pretty confident that things are working as expected (I am running this version on my homebridge server). I just couldn't test the zone control part.

I will leave it for a few days before releasing a new version with the changes to make sure that nothing unexpected shows up.

**Comment by sgraystar** [comment]
Late to the party, thanks greatly for the efforts @yenoiwesa & @TimofeyK 

The unrelease links are not working for me this morning, I imagine you may have further tinkering underway.
I don't have much to add to that above, using an incorrect IP is enough to simulate my original issue.

The additional cases to test would be from the Homekit side, seeing how it reacts when trying/automations/adds/moves/changes during the 85min window.

**Comment by yenoiwesa** [comment]
@sgraystar you should definitely be able to install the unreleased version with:

```
npm install -g yenoiwesa/homebridge-daikin-airbase#8f026c59aa928f95ba4296de2bc3aef6660e671a
```

What error are you getting?

> The additional cases to test would be from the Homekit side, seeing how it reacts when trying automations/adds/moves/changes during the 85min window.

Yes we have both tested that, basically the accessory will appear as "Not responding" in the home app and thus commands will be sent but the plugin rejects them with an error. Siri should tell you "something went wrong" when the airbase hasn't been reattached to the accessory.

**Comment by sgraystar** [comment]
maybe user error?

C:\Users\Touch>npm install -g yenoiwesa/homebridge-daikin-airbase#8f026c59aa928f95ba4296de2bc3aef6660e671a
npm ERR! code ENOENT
npm ERR! syscall spawn git
npm ERR! path git
npm ERR! errno ENOENT
npm ERR! enoent Error while executing:
npm ERR! enoent undefined ls-remote -h -t ssh://git@github.com/yenoiwesa/homebridge-daikin-airbase.git
npm ERR! enoent
npm ERR! enoent
npm ERR! enoent spawn git ENOENT
npm ERR! enoent This is related to npm not being able to find a file.
npm ERR! enoent

npm ERR! A complete log of this run can be found in:
npm ERR!     C:\Users\Touch\AppData\Roaming\npm-cache\_logs\2020-09-01T00_24_14_973Z-debug.log

**Comment by yenoiwesa** [comment]
@sgraystar since you are executing globally make sure you have admin privileges in your terminal. Additionally make sure that you have git installed: `git --version`.
You could also check your npm version `npm --version`

**Comment by sgraystar** [comment]
No git on this machine, I'll look into it

**Comment by yenoiwesa** [comment]
Yeah when pulling straight from the git repository instead of the npm package (where there are only released versions) npm needs git to do that.

**Comment by TimofeyK** [comment]
@sgraystar 
You might need to use the command with sudo depending on how you installed node

**Comment by sgraystar** [comment]
Had a quick look and have not found anything major.

I also use a fixed IP in the config.

vvMinor: Adding the AC accessory to an automation while offline, you cannot change the temperature setpoint from 16 degrees.

When turning the AC on/off while Airbase offline (Automation setup to turn on/off from another light) the same TypeError shows up in the log file:
`[9/1/2020, 11:12:48 AM] [Daikin Airbase] No airbase is associated to Ground Floor
[9/1/2020, 11:13:15 AM] [Daikin Airbase] TypeError: Cannot read property 'ok' of undefined
    at DaikinAircon.sendRequest (C:\Users\Touch\AppData\Roaming\npm\node_modules\homebridge-daikin-airbase\src\airbase-controller.js:128:23)
    at processTicksAndRejections (internal/process/task_queues.js:94:5)
[9/1/2020, 11:13:15 AM] [Daikin Airbase] Will retry to find missing airbase modules with SSIDs [ 'DaikinAP21107' ]`

**Comment by yenoiwesa** [comment]
The `TypeError` is not actually linked to your action in the home app. It's the retry process trying to connect to the wrong IP (it fails to receive a response from that IP hence the error).

The only real error linked to your action is:
`[9/1/2020, 11:12:48 AM] [Daikin Airbase] No airbase is associated to Ground Floor`

> Minor: Adding the AC accessory to an automation while offline, you cannot change the temperature setpoint from 16 degrees.

Yeah I would not have expected any different to be honest. That "offline" mode just throws errors back to homekit when trying to read or set values. It is meant to be a transient state before the accessory comes back to life, so I don't think I should allow users to think they can operate it in that state. The accessory is quite literally not responding.

**Comment by TimofeyK** [comment]
> I could also never drop accessories but then there would be no other way for users than erase entirely the homebridge persistence cache (not just for the Daikin Airbase). That doesn't sound too acceptable

Actually homebridge-config-ui-x plugin allows deleting specific cached accessories from the UI

![F621F7D4-75AC-45D8-8AC5-204B4998D9ED](https://user-images.githubusercontent.com/46880999/91792709-7fe24080-ec59-11ea-94ed-b15882ee306f.jpeg)

**Comment by yenoiwesa** [comment]
Oh wow that's pretty nice @TimofeyK . Maybe it's finally time for me to install that UI... 

Would you then prefer that accessories don't get dropped after the final retry attempt?
I can also change the number of attempts and the exponential factor to make the last attempt happen much later.

**Comment by TimofeyK** [comment]
> Would you then prefer that accessories don't get dropped after the final retry attempt?

Yes. The reason is that removing the physical Airbase controller is a rare and a conscious procedure and it should be expected that it would require some extra steps to remove the relevant Homebridge accessory

Unreachable device (for various reasons such network change, maintenance, power outage etc.) is much more common and removing cached accessory automatically wouldn't be expected and might come as a bad surprise

Not doing any automatic cache removal would simplify the code and prevent such surprises

> Maybe it's finally time for me to install that UI

For me it's an indispensable part of Homebridge! Especially after it introduced hb-service. Greatly simplifies management without adding any overhead. Also Aussie written :-)

**Comment by yenoiwesa** [comment]
I will remove the clean up part then.

> For me it's an indispensable part of Homebridge! Especially after it introduced hb-service. Greatly simplifies management without adding any overhead. Also Aussie written :-)

The problem I have with this plugin is that it requires sudo permission to be useful and has lots of dependencies, many of there can introduce vulnerabilities. I will give it a try without giving it sudo rights see if it's still useful.
I manage my homebridge on a mac mini, using a `.plist` file to auto start homebridge and restart it when it dies.

**Comment by yenoiwesa** [comment]
I have removed the clean up 3e5470ed9ee39bb0f17396ff5ca05d0a7fd42c30.

If we are not expecting any other changes, I will merge the pull request to master and release a new major version tomorrow.

**Comment by yenoiwesa** [comment]
I just released version 3.0.0 of the plugin guys.

You can upgrade to the latest version with:
```
npm install -g homebridge-daikin-airbase@3.0.0
```

**Comment by TimofeyK** [comment]
Will it drop existing accessories/automations for those who upgrading from previous release 2.xx?

If so, then best to mention it at the beginning of the changelog :-)

**Comment by yenoiwesa** [comment]
You're right @TimofeyK, I added a warnin in the release notes [release 3.0.0](https://github.com/yenoiwesa/homebridge-daikin-airbase/releases/tag/3.0.0)

**Comment by yenoiwesa** [comment]
Alright I'll be closing this ticket now. Send a message here if things start failing dramatically :-P
Thanks for your help making the plugin better guys!
